### PR TITLE
Fix CSV uploads for non-ASCII headers

### LINKED ISSUE #47279: Non-ASCII Characters in CSV Column Names Cause SQL Syntax Error in Metabase Upload
### Describe the bug

When uploading a CSV file to Metabase where the column names contain non-ASCII characters (e.g., Japanese characters), the upload fails with an SQL syntax error. This issue occurs despite the fact that rows containing non-ASCII text are handled correctly when column names are in English.

### To Reproduce

1. Create a CSV file with non-ASCII characters (e.g., Japanese) in the column names.
   - Example CSV content:

   ```csv
   ID,名前,年齢,職業,都市
   1,佐藤太郎,25,エンジニア,東京
   2,鈴木花子,30,デザイナー,大阪
   3,田中一郎,28,マーケター,名古屋
   4,山田次郎,35,プロジェクトマネージャー,福岡
   5,中村美咲,32,データサイエンティスト,札幌
   ```

2. Attempt to upload the CSV file to Metabase.

3. Observe the error message:

   ```
   (conn=20363649) You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '%90%8D%E5%89%8D(), E5%B9%B4%E9%BD%A2(), E8%81%B7%E6%A5%AD(), E9%83%BD%E5%B8%82()' at line 1
   ```

### Expected behavior

The CSV file should be uploaded without errors, with non-ASCII characters in column names correctly interpreted and stored.

### Logs

(conn=20363649) You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '%90%8D%E5%89%8D(), E5%B9%B4%E9%BD%A2(), E8%81%B7%E6%A5%AD(), E9%83%BD%E5%B8%82()' at line 1

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-GB",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.6 Safari/605.1.15",
    "vendor": "Apple Computer, Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.23+9",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.23",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.23+9",
    "os.name": "Linux",
    "os.version": "6.1.38-59.109.amzn2023.x86_64",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "mysql",
      "postgres"
    ],
    "hosting-env": "unknown",
    "application-database": "mysql",
    "application-database-details": {
      "database": {
        "name": "MySQL",
        "version": "8.0.37"
      },
      "jdbc-driver": {
        "name": "MariaDB Connector/J",
        "version": "2.7.10"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2024-06-19",
      "tag": "v0.50.6",
      "hash": "a5fbebf"
    },
    "settings": {
      "report-timezone": "Asia/Tokyo"
    }
  }
}
```

### Severity

moderate

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by crisptrutski**
Closes https://github.com/metabase/metabase/issues/47279### 

### Description

It turns out that we had a chink in our support for non-ASCII characters - column names that start with a non-ASCII character.

This is a bit subtle, and due to the combination of our munging technique and how HoneySQL handles function calls.

- For munging, we use a shared helper that uses URL encoding internally, which prefixes escape sequences with `%`
- HoneySQL interprets keywords starting with `%` as function calls (with no argument).

~~The fix is pretty straight forward, if you namespace a field to qualify it, HoneySQL then knows it is a field.~~ For some reason this did not work for MySQL based engines, so I found another solution - manually formatting the references and putting them in `:raw` blocks.

### CODE REVIEW

File: src/metabase/upload.clj [hunk 1]
```diff
@@ -68,6 +69,29 @@
   (let [column-limit (some-> driver driver/column-name-length-limit)]
     (min-safe column-limit (max-bytes :model/Field :name))))
 
+(defn truncate-utf8-bytes
```
Comment (crisptrutski) [inline src/metabase/upload.clj:72]: This really sucks, I'm surprised there is not something built in. Maybe need to look harder?

Comment (crisptrutski) [reply src/metabase/upload.clj]: Pushed this to a later PR

File: test/metabase/upload_test.clj [hunk 2]
```diff
@@ -744,6 +743,34 @@
...
+                                       "3,田中一郎,28,マーケター,名古屋"
+                                       "4,山田次郎,35,プロジェクトマネージャー,福岡"
+                                       "5,中村美咲,32,データサイエンティスト,札幌"]))]
+         (testing "Check the data was uploaded into the table correctly"
+           (is (= #_(header-with-auto-pk ["ID" "名前" "年齢" "職業" "都市"])
+                  (header-with-auto-pk ["id"
+                                        "%e5%90%8d%e5%89%8d"
+                                        "%e5%b9%b4%e9%bd%a2"
+                                        "%e8%81%b7%e6%a5%ad"
+                                        "%e9%83%bd%e5%b8%82"])
```
Comment (crisptrutski) [inline test/metabase/upload_test.clj:764]: Will improve this in the next PR

File: test/metabase/upload_test.clj [hunk 3]
```diff
@@ -2265,6 +2292,6 @@
 (defmethod driver/column-name-length-limit ::short-column-test-driver [_] 10)
 
 (deftest unique-long-column-names-test
-  (let [original ["αbcdεf"     "αbcdεfg"   "αbc_2_etc" "αbc_3_xyz"]
+  (let [original ["αbcdεf_αbcdεf"     "αbcdεfg_αbcdεf"   "αbc_2_etc_αbcdεf" "αbc_3_xyz_αbcdεf"]
```
Comment (crisptrutski) [inline test/metabase/upload_test.clj:2295]: Whoops, accidentally spilled in from next change. Needed more characters to test truncation of display names.

No harm leaving it in.

File: src/metabase/driver/sql_jdbc.clj [hunk 4]
```diff
@@ -170,8 +168,9 @@
         ;; across all drivers. With that in mind, 100 seems like a safe compromise.
         ;; There's nothing magic about 100, but it felt good in testing. There could well be a better number.
         chunks     (partition-all (or driver/*insert-chunk-rows* 100) values)
-        sqls       (map #(sql/format {:insert-into table-name
-                                      :columns     columns
+        sqls       (map #(sql/format {:insert-into (keyword table-name)
+                                      ;; We need to namespace the keyword in case the column name starts with a %
+                                      :columns     (map (comp (partial keyword table-name) name) column-names)
```
Comment (piranha) [inline src/metabase/driver/sql_jdbc.clj:173]: uhhh that `comp` (and the one in postgres.clj) is way harder to read compared to the `#(...)` version in mysql.clj

**Review by sansolovyov — APPROVED** [review]
👌 except I'd love to see uniformity in how `:columns` are generated

**Pull request merged**